pkg/manager/member/pvc_cleaner.go

```diff
@@ -40,37 +49,160 @@ type PVCCleanerInterface interface {
 }
 
 type realPVCCleaner struct {
+	kubeCli    kubernetes.Interface
 	podLister  corelisters.PodLister
 	pvcControl controller.PVCControlInterface
 	pvcLister  corelisters.PersistentVolumeClaimLister
+	pvLister   corelisters.PersistentVolumeLister
+	pvControl  controller.PVControlInterface
 }
 
 // NewRealPVCCleaner returns a realPVCCleaner
 func NewRealPVCCleaner(
+	kubeCli kubernetes.Interface,
 	podLister corelisters.PodLister,
 	pvcControl controller.PVCControlInterface,
-	pvcLister corelisters.PersistentVolumeClaimLister) PVCCleanerInterface {
+	pvcLister corelisters.PersistentVolumeClaimLister,
+	pvLister corelisters.PersistentVolumeLister,
+	pvControl controller.PVControlInterface) PVCCleanerInterface {
 	return &realPVCCleaner{
+		kubeCli,
 		podLister,
 		pvcControl,
 		pvcLister,
+		pvLister,
+		pvControl,
 	}
 }
 
 func (rpc *realPVCCleaner) Clean(tc *v1alpha1.TidbCluster) (map[string]string, error) {
+	if skipReason, err := rpc.cleanScheduleLock(tc); err != nil {
+		return skipReason, err
+	}
+
+	if !tc.Spec.EnablePVReclaim {
+		// disable PV reclaim, return directly.
+		return nil, nil
+	}
+
+	return rpc.reclaimPV(tc)
+}
+
+func (rpc *realPVCCleaner) reclaimPV(tc *v1alpha1.TidbCluster) (map[string]string, error) {
 	ns := tc.GetNamespace()
 	tcName := tc.GetName()
+
 	// for unit test
 	skipReason := map[string]string{}
 
-	selector, err := label.New().Instance(tc.GetLabels()[label.InstanceLabelKey]).Selector()
+	pvcs, err := rpc.listAllPVCs(tc)
 	if err != nil {
-		return skipReason, fmt.Errorf("cluster %s/%s assemble label selector failed, err: %v", ns, tcName, err)
+		return skipReason, err
 	}
 
-	pvcs, err := rpc.pvcLister.PersistentVolumeClaims(ns).List(selector)
+	for _, pvc := range pvcs {
+		pvcName := pvc.GetName()
+		l := label.Label(pvc.Labels)
+		if !(l.IsPD() || l.IsTiKV()) {
+			skipReason[pvcName] = skipReasonPVCCleanerIsNotPDOrTiKV
+			continue
+		}
+
+		if pvc.Status.Phase != corev1.ClaimBound {
+			// If pvc is not bound yet, it will not be processed
+			skipReason[pvcName] = skipReasonPVCCleanerPVCNotBound
+			continue
+		}
+
+		if pvc.DeletionTimestamp != nil {
+			// PVC has been deleted, skip it
+			skipReason[pvcName] = skipReasonPVCCleanerPVCHasBeenDeleted
+			continue
+		}
+
+		if len(pvc.Annotations[label.AnnPVCDeferDeleting]) == 0 {
```

Comment:
Suggest to check nil for pvc.Annotations.

Reply:
`len(pvc.Annotations[label.AnnPVCDeferDeleting]) == 0` including this situation.